---
title: "MainMarkdown"
output: html_document
---

```{r load dependencies, echo = F}
library(tidyverse)
library(limma)
library(Biobase)
library(biomaRt)
```

```{r setup, include = F}
gxData <- read.delim("./Data/MAGNET_GeneExpressionData_CPM_19112020.txt", as.is = T, row.names = 1)
sampleInfo <- read.delim("./Data/MAGNET_SampleData_19112020.txt", as.is = T, row.names = 1)
geneTotExonLengths <- read.delim("./Data/MAGNET_exonLengths.txt", as.is = T, row.names = 1)
```


```{r convert cpm to fpkm values}
all(rownames(geneTotExonLengths) == rownames(gxData)) # TRUE (just a check)
cpm2fpkm <- function(x) {
	geneTotExonLengths_kb <- geneTotExonLengths[, 1] / 1E3
	.t <- 2^(x) / geneTotExonLengths_kb
#	.t <- 2^(x) * 1E3 / geneTotExonLengths[, 1] # this does the same, but shorter
}
gxData_fpkm <- cpm2fpkm(gxData)
```

differential gene expression
```{r Differential Gene Expression}
# Create ExpressionSet object
DifferentialGeneExpressionCalc <- function(assayData, phenoData){
  # Create ExpressionSet object
  eset <- ExpressionSet(assayData = data.matrix(assayData), phenoData = AnnotatedDataFrame(phenoData))
  design <- model.matrix(~0+Disease, data = pData(eset))

  df <- data.frame(matrix(ncol = 0, nrow = length(as.matrix(assayData[1]))))

  
  # print(design[,diseasDesignCol])
  # statusTest <-
  # print(design[,"DiseaseDonor"])
  # Create a contrasts matrix
  cm <- makeContrasts(
  DonorVsPPCM = DiseaseDonor - DiseasePPCM,
  DonorVsHCM = DiseaseDonor - DiseaseHCM,
  DonorVsDCM = DiseaseDonor - DiseaseDCM,
  levels = design)
  
  for(comparison in colnames(cm)){
    # Fit the model
    fit <- lmFit(eset, design)
    # Fit the contrasts
    fit2 <- contrasts.fit(fit, contrasts = cm[,comparison])
    # Calculate the t-statistics for the contrasts
    fit2 <- eBayes(fit2)

    diffGenExprTestData <- topTable(fit2, adjust = "fdr", number = nrow(gxData))
    diffGenExprTestDecision <- decideTests(fit2, adjust.method = "fdr", p.value = 0.05, lfc = 1)
    colnames(diffGenExprTestData) <- paste(colnames(diffGenExprTestData), comparison, sep = "_")
    colnames(diffGenExprTestDecision) <- paste(colnames(diffGenExprTestDecision), comparison, sep = "_")

    df <- cbind(df, diffGenExprTestData, diffGenExprTestDecision)
  }
  df
}
test <- DifferentialGeneExpressionCalc(gxData, sampleInfo)

#remove objects that are no longer needed
rm(eset, design, cm, fit, fit2)
  
```

```{r biomart gene annotation}

mart <- useMart("ensembl")
#set to use ensemble database for humans
mart <- useDataset("hsapiens_gene_ensembl",mart)
# listAttributes(mart) #to get a list of possible attributes

#get gene annotation data from ensembl database
geneAnnotationData <- getBM(attributes = c("ensembl_gene_id", "go_id", "entrezgene_id", "chromosome_name", "external_gene_name", "hgnc_symbol", "description"), 
                            mart = mart,
                            filters = "ensembl_gene_id",
                            values = rownames(gxData)
                            )
```

```{r Gene Enrichment Analysis, echo = F}



```
